runtime/lua/vim/treesitter/languagetree.lua

```diff
@@ -409,6 +428,45 @@ function LanguageTree:_add_injections()
   return query_time
 end
 
+--- Run an asynchronous parse, calling {on_parse} when complete.
+---
+--- @param range boolean|Range|nil
+--- @param on_parse fun(trees: TSTrees, timeout_expired: boolean)
+function LanguageTree:_async_parse(range, on_parse)
+  coroutine.resume(coroutine.create(function()
+    local co = coroutine.running()
+    local ct = self:_buf().changedtick
+    local total_parse_time = 0
+    local redrawtime = vim.o.redrawtime
+    ---@type TSTrees, boolean
+    local trees, unfinished
+
+    while total_parse_time <= redrawtime do
+      -- If buffer was changed in the middle of parsing, reset parse state
+      if self:_buf().changedtick ~= ct then
+        self._parser:reset()
+        self:invalidate()
+        total_parse_time = 0
+        coroutine.yield()
+      end
+
+      trees, unfinished = self:_parse(range, default_parse_timeout_ms)
+      total_parse_time = total_parse_time + default_parse_timeout_ms
```

Comment:
We should add the time it actually took to parse rather than adding the timeout value. We can use `tcall()` for that.

Reply:
I had a question about this actually, in [this comment](https://github.com/neovim/neovim/pull/31631#discussion_r1894996407)